nvflare/fuel/utils/hash_utils.py

```diff
@@ -0,0 +1,52 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import hashlib
+
+from nvflare.fuel.utils.validation_utils import check_number_range, check_str
+
+# A large prime number as virtual hash table size
+PRIME = 100003
+MAX_NUM_BUCKETS = 64 * 1024
+
+
+class UniformHash:
+    """A hash algorithm with uniform distribution. It achieves this with following steps,
+    1. Get a hash value using SHA256
+    2. Map the hash value to a virtual hash table
+    3. Map the virtual bucket to real bucket
+
+    """
+
+    def __init__(self, num_buckets: int):
+        check_number_range("num_buckets", num_buckets, 1, MAX_NUM_BUCKETS)
+        self.num_buckets = num_buckets
+        self.virtual_hashes_per_bucket = PRIME // num_buckets
+
+    def get_num_buckets(self) -> int:
+        return self.num_buckets
+
+    def hash(self, key: str) -> int:
+        check_str("key", key)
+        # The hash() function changes value every run so SHA256 is used
+        sha_bytes = hashlib.sha256(key.encode()).digest()
+        sha = int.from_bytes(sha_bytes[:8], "big")
```

Comment:
I remember that in your previous version, you used crc to compute the hash key. Why did you change?
I also tried zlib.crc32(key.encode(), 0), and compared standard deviation between sha and crc-based methods for randomly generated list of keys. CRC's stdev is smaller/better most of the times (8 of 10 times) than sha-based.

Reply:
Actually their performance is about the same if trying enough times.
But CRC uses about half compute time.
I guess either is okay.
Another finding is that if you simply do int.from_bytes(key.encode[:8], "big"), it also works equally well, but takes less time. Of course the key must have enough 8 bytes.

Reply:
Since we are making this part of util, I selected SHA256 for its robustness. CRC32 is not considered a real hash algorithm. We are not very sensitive to performance so this is ok.